mlir/lib/Dialect/MQTOpt/Transforms/GateDecompositionPattern.cpp

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (c) 2023 - 2025 Chair for Design Automation, TUM
+ * Copyright (c) 2025 Munich Quantum Software Company GmbH
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * Licensed under the MIT License
+ */
+
+#include "Helpers.h"
+#include "mlir/Dialect/MQTOpt/IR/MQTOptDialect.h"
+#include "mlir/Dialect/MQTOpt/Transforms/Passes.h"
+
+#include <iterator>
+#include <llvm/ADT/STLExtras.h>
+#include <mlir/Dialect/Arith/IR/Arith.h>
+#include <mlir/IR/MLIRContext.h>
+#include <mlir/IR/Operation.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/IR/ValueRange.h>
+#include <mlir/Support/LLVM.h>
+#include <mlir/Support/LogicalResult.h>
+
+namespace mqt::ir::opt {
+/**
+ * @brief This pattern TODO.
+ */
+struct EulerDecompositionPattern final
+    : mlir::OpInterfaceRewritePattern<UnitaryInterface> {
+
+  explicit EulerDecompositionPattern(mlir::MLIRContext* context)
+      : OpInterfaceRewritePattern(context) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(UnitaryInterface op,
+                  mlir::PatternRewriter& rewriter) const override {
+    if (!helpers::isSingleQubitOperation(op)) {
+      return mlir::failure();
+    }
+
+    auto series = getSingleQubitSeries(op);
+    if (series.size() <= 3) {
+      // TODO: find better way to prevent endless optimization loop
+      return mlir::failure();
+    }
+
+    dd::GateMatrix unitaryMatrix = dd::opToSingleQubitGateMatrix(qc::I);
```

Comment:
We should not rely on the `dd::GateMatrix` representation. MLIR/LLVM should have builtin support for working with matrices (or tensors) and their manipulations.
Maybe https://mlir.llvm.org/docs/Dialects/Linalg/ is a good first stop or https://www.stephendiehl.com/posts/mlir_linear_algebra/.

Generally, I would argue that we should be taking advantage of MLIR's infrastructure as much as possible.